metadata_rules/handler.py

```diff
@@ -0,0 +1,68 @@
+# handler.py
+""" Module to launch application """
+
+import _set_path  # noqa
+import os
+import json
+from pathlib import Path
+from pipelineutilities.pipeline_config import setup_pipeline_config, load_config_ssm  # noqa: E402
+import sentry_sdk  # noqa: E402
+from sentry_sdk.integrations.aws_lambda import AwsLambdaIntegration  # noqa: E402
+from harvest_metadata_rules import HarvestMetadataRules
+from dependencies.sentry_sdk import capture_exception
+from distutils.dir_util import copy_tree
+
+
+if 'SENTRY_DSN' in os.environ:
+    sentry_sdk.init(dsn=os.environ['SENTRY_DSN'], integrations=[AwsLambdaIntegration()])
+
+
+def run(event, context):
+    """ run the process to retrieve and process web kiosk metadata """
+    _suplement_event(event)
+    config = setup_pipeline_config(event)
+    google_config = load_config_ssm(config['google_keys_ssm_base'])
+    config.update(google_config)
+    google_credentials = json.loads(config["museum-google-credentials"])
+    harvest_metadata_rules_class = HarvestMetadataRules(google_credentials)
+    local_folder = os.path.dirname(os.path.realpath(__file__)) + '/'
+    for site_name in event['sites']:
+        harvest_metadata_rules_class.harvest_google_spreadsheet_info(site_name)
+    _save_sites_info_to_s3(local_folder + "sites/", "s3://" + config["process-bucket"] + "/sites/")
+    try:
+        copy_tree(local_folder + "sites/", local_folder + "../process_manifest/sites/")
+    except Exception:
+        print('Unable to sync sites files to process_manifest ')
+        pass
+
+
+def _save_sites_info_to_s3(sites_folder, s3_key):
+    try:
+        sync_command = f"aws s3 sync " + sites_folder + " " + s3_key
+        os.system(sync_command)
+    except Exception as e:
+        print('Unable to sync sites files (' + sites_folder + ') to the process bucket here:', s3_key)
+        capture_exception(e)
```

Comment:
I understand that boto3 doesn't have something out of the box to sync but doesn't seem terrible to write. Seems like a good utility method for pipeline_utilities/s3_helpers.py. See here - https://stackoverflow.com/questions/53517519/sync-two-buckets-through-boto3
Maybe that'd take too to sync via code idk. Just a thought.


If that's the case I wouldn't recommend os.system. If s3_key were some malicious command('; rm -rf *') os.system would execute it. Instead use subprocess - https://docs.python.org/3/library/subprocess.html#replacing-os-system

Reply:
@rdoughty the write_s3_* methods in our s3_helpers implement the s3 etag calculation on data being sent to s3 and can tell if they should copy the file or not.   If something other than boto is used i would use those methods.